File: app/src/main/java/me/instabattle/app/managers/EntryManager.java
package me.instabattle.app.managers;

import com.google.android.gms.maps.model.LatLng;

import java.util.List;

import me.instabattle.app.models.Entry;
import me.instabattle.app.models.Vote;
import me.instabattle.app.services.LocationService;
import me.instabattle.app.settings.State;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;

public class EntryManager {
    private static final EntryService service = ServiceGenerator.createService(EntryService.class);
    private static EntryService tokenService;

    public static void initTokenService() {
        tokenService = ServiceGenerator.createService(EntryService.class, State.token);
    }

    public static void getByBattleAndDo(int battleId, Callback<List<Entry>> callback) {
        Call<List<Entry>> call = service.getByBattle(battleId);
        call.enqueue(callback);
    }

    public static void getTopByBattleAndDo(int battleId, int count, Callback<List<Entry>> callback) {
        Call<List<Entry>> call = service.getTopByBattle(battleId, count);
        call.enqueue(callback);
    }

    public static void getByUserAndDo(int userId, Callback<List<Entry>> callback) {
        Call<List<Entry>> call = service.getByUser(userId);
        call.enqueue(callback);
    }

    public static void getAndDo(Integer entryId, Callback<Entry> callback) {
        Call<Entry> call = service.get(entryId);
        call.enqueue(callback);
    }

    public static void createAndDo(Integer battleId, Integer userId, Callback<Entry> callback) {
        LatLng loc = LocationService.getCurrentLocation();
        Call<Entry> call = tokenService.create(new Entry(battleId, userId, loc.latitude, loc.longitude));
        call.enqueue(callback);
    }

    public static void getVoteAndDo(Integer battleId, Callback<List<Entry>> callback) {
        Call<List<Entry>> call = service.getVote(battleId, State.currentUser.getId());
        call.enqueue(callback);
    }

    public static void voteAndDo(Integer battleId, Integer voterId, Integer winnerId, Integer loserId,
                                 Callback<Vote> callback) {
        Call<Vote> call = tokenService.vote(battleId, new Vote(voterId, winnerId, loserId));
        call.enqueue(callback);
    }

    interface EntryService {
        @GET("entries/{entry_id}")
        Call<Entry> get(@Path("entry_id") Integer entryId);

        @GET("users/{user_id}/entries")
        Call<List<Entry>> getByUser(@Path("user_id") Integer userId);

        @GET("battles/{battle_id}/entries")
        Call<List<Entry>> getByBattle(@Path("battle_id") Integer battleId);

        @GET("battles/{battle_id}/entries")
        Call<List<Entry>> getTopByBattle(@Path("battle_id") Integer battleId,
                                         @Query("count") Integer count);

        @GET("battles/{battle_id}/voting")
        Call<List<Entry>> getVote(@Path("battle_id") Integer battleId,
                                  @Query("user_id") Integer userId);

        @POST("battles/{battle_id}/voting")
        Call<Vote> vote(@Path("battle_id") Integer battleId,
                               @Body Vote vote);

        @POST("entries/")
        Call<Entry> create(@Body Entry entry);
    }
}

File: app/src/main/kotlin/me/instabattle/app/managers/EntryManager.kt
package me.instabattle.app.managers

import me.instabattle.app.models.Entry
import me.instabattle.app.models.Vote
import me.instabattle.app.services.LocationService
import me.instabattle.app.settings.State
import retrofit2.Call
import retrofit2.Callback
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

object EntryManager {
    private val service = ServiceGenerator.createService(EntryService::class.java)
    private var tokenService: EntryService? = null

    fun initTokenService() {
        tokenService = ServiceGenerator.createService(EntryService::class.java, State.token)
    }

    fun getByBattleAndDo(battleId: Int, callback: Callback<List<Entry>>) {
        val call = service.getByBattle(battleId)
        call.enqueue(callback)
    }

    fun getTopByBattleAndDo(battleId: Int, count: Int, callback: Callback<List<Entry>>) {
        val call = service.getTopByBattle(battleId, count)
        call.enqueue(callback)
    }

    fun getByUserAndDo(userId: Int, callback: Callback<List<Entry>>) {
        val call = service.getByUser(userId)
        call.enqueue(callback)
    }

    fun getAndDo(entryId: Int?, callback: Callback<Entry>) {
        val call = service[entryId]
        call.enqueue(callback)
    }

    fun createAndDo(battleId: Int?, userId: Int?, callback: Callback<Entry>) {
        val loc = LocationService.getCurrentLocation()
        val call = tokenService!!.create(Entry(battleId!!, userId!!, loc.latitude, loc.longitude))
        call.enqueue(callback)
    }

    fun getVoteAndDo(battleId: Int?, callback: Callback<List<Entry>>) {
        val call = service.getVote(battleId, State.currentUser.id)
        call.enqueue(callback)
    }

    fun voteAndDo(battleId: Int?, voterId: Int?, winnerId: Int?, loserId: Int?,
                  callback: Callback<Vote>) {
        val call = tokenService!!.vote(battleId, Vote(voterId, winnerId, loserId))
        call.enqueue(callback)
    }

    internal interface EntryService {
        @GET("entries/{entry_id}")
        operator fun get(@Path("entry_id") entryId: Int?): Call<Entry>

        @GET("users/{user_id}/entries")
        fun getByUser(@Path("user_id") userId: Int?): Call<List<Entry>>

        @GET("battles/{battle_id}/entries")
        fun getByBattle(@Path("battle_id") battleId: Int?): Call<List<Entry>>

        @GET("battles/{battle_id}/entries")
        fun getTopByBattle(@Path("battle_id") battleId: Int?,
                           @Query("count") count: Int?): Call<List<Entry>>

        @GET("battles/{battle_id}/voting")
        fun getVote(@Path("battle_id") battleId: Int?,
                    @Query("user_id") userId: Int?): Call<List<Entry>>

        @POST("battles/{battle_id}/voting")
        fun vote(@Path("battle_id") battleId: Int?,
                 @Body vote: Vote): Call<Vote>

        @POST("entries/")
        fun create(@Body entry: Entry): Call<Entry>
    }
}
